sdks/java/extensions/ml/src/main/java/org/apache/beam/sdk/extensions/ml/RecommendationAIImportUserEvents.java

```diff
@@ -0,0 +1,216 @@
+ * to "default_catalog"). It is possible to provide a event store to which you want to add the user
+ * event (defaults to "default_event_store").
+ */
+@AutoValue
+@SuppressWarnings({"nullness"})
+public abstract class RecommendationAIImportUserEvents
+    extends PTransform<PCollection<KV<String, GenericJson>>, PCollectionTuple> {
+
+  public static final TupleTag<UserEvent> SUCCESS_TAG = new TupleTag<UserEvent>() {};
+  public static final TupleTag<UserEvent> FAILURE_TAG = new TupleTag<UserEvent>() {};
+
+  public static Builder newBuilder() {
+    return new AutoValue_RecommendationAIImportUserEvents.Builder()
+        .setCatalogName("default_catalog")
+        .setEventStore("default_event_store");
+  }
+
+  public abstract Builder toBuilder();
+
+  /** @return ID of Google Cloud project to be used for creating user events. */
+  public abstract @Nullable String projectId();
+
+  /** @return Name of the catalog where the user events will be created. */
+  public abstract @Nullable String catalogName();
+
+  /** @return Name of the event store where the user events will be created. */
+  public abstract @Nullable String eventStore();
+
+  /** @return Size of input elements batch to be sent in one request. */
+  public abstract Integer batchSize();
+
+  /**
+   * @return Time limit (in processing time) on how long an incomplete batch of elements is allowed
+   *     to be buffered.
+   */
+  public abstract Duration maxBufferingDuration();
+
+  public RecommendationAIImportUserEvents withProjectId(String projectId) {
+    return this.toBuilder().setProjectId(projectId).build();
+  }
+
+  public RecommendationAIImportUserEvents withCatalogName(String catalogName) {
+    return this.toBuilder().setCatalogName(catalogName).build();
+  }
+
+  public RecommendationAIImportUserEvents withEventStore(String eventStore) {
+    return this.toBuilder().setEventStore(eventStore).build();
+  }
+
+  public RecommendationAIImportUserEvents withBatchSize(Integer batchSize) {
+    return this.toBuilder().setBatchSize(batchSize).build();
+  }
+
+  /**
+   * The transform converts the contents of input PCollection into {@link UserEvent}s and then calls
+   * the Recommendation AI service to create the user event.
+   *
+   * @param input input PCollection
+   * @return PCollection after transformations
+   */
+  @Override
+  public PCollectionTuple expand(PCollection<KV<String, GenericJson>> input) {
+    return input
+        .apply(
+            "Batch Contents",
+            GroupIntoBatches.<String, GenericJson>ofSize(batchSize())
+                .withMaxBufferingDuration(maxBufferingDuration())
+                .withShardedKey())
+        .apply(
+            "Import CatalogItems",
+            ParDo.of(new ImportUserEvents(projectId(), catalogName(), eventStore()))
+                .withOutputTags(SUCCESS_TAG, TupleTagList.of(FAILURE_TAG)));
+  }
+
+  @AutoValue.Builder
+  public abstract static class Builder {
```

Comment:
you can remove `public` from this builder so it can only be used within the package